modules/renter/contractor/journal.go

```diff
@@ -105,10 +107,18 @@ func (j *journal) checkpoint(data contractorPersist) error {
 	if err := tmp.Close(); err != nil {
 		return err
 	}
-	if err := j.f.Close(); err != nil {
-		return err
+
+	// If rename or close failed in a previous call, fd will be invalid,
+	//	so don't try to close
+	if fd := j.f.Fd(); fd != uintptr(syscall.InvalidHandle) {
+		if err := j.f.Close(); err != nil {
+			return err
```

Comment:
shouldn't `Close` always work unless `j.f == nil`?
I think we can actually ignore the error here, since we call `Sync` beforehand. i.e.
```go
if j.f. != nil {
    j.f.Close()
}
```

Reply:
I guess it depends on what should be logged in consensus.  Sorry if syntax is incorrect. f.n is never nil, just the fd is set to 'syscall.InvalidHandle'.
- Option 1 - Ignore Close error altogether
```
j.f.Close()
if err := os.Rename(tmp.Name(), j.filename); err != nil {
   return err;
}
```
Invalid handle will not be logged, but rename will log 'Access is denied', which could be misleading as the root cause is invalid handle.


- Option 2 - Only rename if Close() is successful
```
if err := j.f.Close(); err == nil {
      if err := os.Rename(tmp.Name(), j.filename); err != nil {
       return err;
    }
}
```
Only rename errors are logged, but have no way of knowing there's an issue if Close() returns an error


- Option 3 - Log root cause if either Close() or Rename() fail
```
if err := j.f.Close(); err != nil {
   return err;
} else if err := os.Rename(tmp.Name(), j.filename); err != nil {
       return err;
    }
}
```

Reply:
Do we know *why* the fd gets set to `InvalidHandle`? Somehow antivirus is interfering? Dropping down to the `syscall` interface is more low-level than I'd prefer. Maybe if we can figure out the root cause, we can handle the problem at a higher level.

Why does `Rename` return `Access is denied` if the handle is invalid?

Reply:
It becomes invalid because the rename fails and the reopen isn't called. The next call starts with a closed handle while expecting an open one. Reopen is never being called again when rename failed.

Reply:
close sets the handle to invalid on success.

Reply:
as for access is denied, looks like that's what is returned by go's file system calls when trying to rename given an invalid fd.

Reply:
essentially, reopen has to be called if either close or rename fails. otherwise, stuck in an invalid fd loop.

Reply:
stuck because close returns an error when trying to close an invalid handle.

Reply:
so in principle, this shouldn't happen as long as the `Rename` (eventually) succeeds? Maybe the addition of `build.Retry` will be enough to prevent the `InvalidHandle` bug. If the `Rename` can't succeed after 5 retries, that's probably sufficient cause to crash the program. We definitely don't want it to be silently failing in the background.

Reply:
I would suspect drive failure at that point or some other process has opened a handle. contractor.log fills to tens of megabytes if i leave to it's own devices. siad does eventually execute an orderly shutdown in >5 minutes or so.

Reply:
if rename succeeds, it's fine. out of 3 win instances only one with McAfee AV ever gets into this state.

Reply:
ok. My preference is to treat the `j.f.Close()` error normally (i.e. how it was treated before the PR), and if the `build.Retry` error is not nil, crash the program using `build.Critical("Unable to save contractor.log:", err)`. Are you able to test that on your McAfee machine? How reproducible is this?

Reply:
and actually rename could fail as long as the failure isn't due to invalid fd. root of it is closing an invalid handle. if it's valid, the next call will simply close it and try the rename again.

Reply:
immediately reproducible. it never launches again.